### ISSUE TITLE
Enhance UI for templates

### ISSUE THREAD

**Issue by KayRJay** [ericberman/MyFlightbookWeb#478]
The current user interface for creating, editing and deleting templates is incomplete and a bit awkward (and different from the normal logbook view).   Editing a template is done by clicking on the pencil, on the left, while deleting a template is done by clicking on the X on the right.

Creating a new template requires starting from scratch.  Since many templates have quite a few properties, it would be helpful to be able to copy a template, and then rename it.

As with flights on the logbook page, operations on templates could be chosen from a drop-down menu on the right side:
<img width="274" alt="Screen Shot 2020-02-01 at 10 45 05 AM" src="https://user-images.githubusercontent.com/16978320/73596580-77b5fd00-44e0-11ea-9fb8-4e9193b90987.png">
When a template is copied, it would be added to the user's templates with the name _Template name copy_ and the user could subsequently rename it.  As with the macOS Finder, multiple copies would be numbered: _Template name copy 2_, etc.

With this approach, there needs to be some indication of those templates that are shared or default.   Icons on the left of the template name would work well:
<img width="699" alt="Screen Shot 2020-02-01 at 11 27 53 AM" src="https://user-images.githubusercontent.com/16978320/73597109-f3ff0f00-44e5-11ea-9128-8cee1eefaff4.png">
The drop-down choices to share or use a template as default would toggle that status and add or remove the corresponding icon.

By the way, it would be helpful if the template page didn't lose the scrolling position after the user clicks one of the choices to edit, delete, etc. a template.

Further, the group containing templates might better be labelled "Flight Properties and Templates".

Also, while the user can go back to [Browse Available Templates](https://myflightbook.com/logbook/Member/BrowseTemplates.aspx), it takes 3 or 4 clicks after browsing to continue working with templates that are in the user profile.   There should be a link on the browsing page, "Review my templates".

Finally, rather than have the text "Add to my templates" or "Template added to your account" on the right side of the template browsing page, it might be cleaner to have a single check box to indicate whether a template is or is not  in the user's profile (or is "in my templates" or is "in my account").   The check itself would not be clickable.  The semantic behavior would remain as is, with the override check if the user attempts to re-add a template to his profile.  Just replace the text with a heading and a single icon.
<img width="503" alt="Screen Shot 2020-02-01 at 11 55 10 AM" src="https://user-images.githubusercontent.com/16978320/73597486-bbf9cb00-44e9-11ea-9252-832c627087f2.png">

**Comment by ericberman** [comment]
Good ideas.  I think the drop-down is an interesting idea.  I use the pencil/x paradigm in a lot of places throughout the site (though I don't think I'm super consistent about which is on the left and which is on the right.)

I like the copy idea.  That's part of the point of having the shared community library - when you add from that library the semantics are precisely "make a copy of this and put it into my profile".  But my current model doesn't let you copy from yourself, which your suggestion does.

Having the "public" and "Default" toggles indicated on the left and modified on the drop-down is a bit of a weird dichotomy.  You correctly recognize that unlike the other three options (Edit/delete/copy), Share and Use by Default have both a state-indication requirement and a change-the-state requirement.  I went with checkboxes because they neatly ecapsulate both - indeed, that's precisely the purpose of checkboxes.  So the real tradeoff is that you can put more stuff into the drop-down at the expense of losing the visible state indication, which means you need to bring back the state indication (as you propose with the icons), which eats up the space you just saved - i.e., at that point you haven't really solved anything except replaced a column that is wide due to the header to one that is narrow because the icon communicates the header.  I wonder if there's some way to do that in a way that is clear.  I'd still want the icons to be clickable.

I think there was a reason I couldn't do edits in an update panel (both reduces flashing and preserves scroll position).  Might have been that I don't do the edit in-place (I populate the already displayed editing UI), but I can play with that.

**Comment by KayRJay** [comment]
We are on the same page with this.

**Comment by ericberman** [comment]
I am going to leave the checkboxes and headers.  I don't think an icon is sufficient for what are somewhat abstract (non-visual) ideas as "shared" and "default", so I need the header.  I've rotated it 45 degrees, though, for space.  But I've done pretty much everything else here.  Should be live in the next few days.

**Comment by KayRJay** [comment]
I look forward to trying these new features.   

As noted in issue #479, if a template is to be used to control the logbook view format, the definition of the template would need to include the list of property values to be displayed and (secondarily) their left-to-right order.    I make this comment here so we can revisit the related UI visits as part of this Github issue.

**Comment by ericberman** [comment]
I think these (property templates) and a logbook template are different and independent things.  You change property template from flight-to-flight.  E.g., I already have automatic templates that come and go based on whether an aircraft is anonymous (in order to capture a tailnumber, if you like) or a sim (to capture serial # and any ground-instruction time).  Or, you might be an airline pilot during the week and fly gliders on the weekend.  

In other words, property templates control the entry of a single specific flight; logbook templates (in 479) control how a collection of heterogeneous flights are displayed).  They're independent and distinct things.

**Comment by KayRJay** [comment]
Ok, that makes sense.   So, on the template UI, you'd create/edit EITHER a data entry template or a logbook template. 

Both types of templates are about properties.  The terminology for the data entry template should probably be something other than "property template".   "Data entry template" is pretty straightforward and unambiguous.

**Comment by KayRJay** [comment]
Both types of templates could be used to define ordering, as mentioned in issue #479.

The order is probably more important for the left-to-right order of display in the logbook than for data entry, but it seems reasonable to apply the ordering in both cases.

**Comment by ericberman** [comment]
I hadn't been thinking that the logbook template is about properties.  yes, it overlaps in that you could add/remove various columns, some of which would correspond to properties.  But I was thinking of the logbook display template as being about layout as well - what order things are, what is shown, and so forth: are pictures included?  Instructor signatures?  Engine times and flight times (already an option)?  Routes?  (Note that all of those currently all currently live in a single column).  And there are computed properties (Do you compute day flight, which is currently not stored anywhere, and display that?), and issues of redundancy (if you put Route into its own column, you probably want to take it out of the first column).  

I.e., can you make them look distinct like the printing templates currently do?

If it's just a matter of being able to remove or add a few columns of data, then indeed logbook templates do become a lot simpler.  That's not super hard at all.

**Comment by ericberman** [comment]
By the way, Property templates specify *which* properties are shown for data entry, but do not say anything about order.  They're sorted (mostly) alphabetically.

**Comment by KayRJay** [comment]
Some interesting thoughts about the **potential** for logbook templates.   However, this is probably an area where less is more.   The various things you mention are (or could/should be) available on the Details view for a flight.  There's not a real need to control the entire layout of the logbook view with a template, nor to display derivative values or most of the other things you mention.   It seems sufficient to me to simply control what basic properties are shown in what order.

Re the examples you mentioned of possibilities, here's my take:

- Pictures ... pretty disruptive to the (basically) line-by-line format.   Can you do what you do for aircraft pictures (cursor hover over the tail number)?   Add a little icon in the flight row to indicate a picture is available.
- Can you (redundantly?) store such things as instructor signatures, engine times and flight times  and even routes as properties?  Maybe the UI doesn't even need to change, just change the underlying data representation, and make such new properties available for templates?
- Computed properties ... seems unnecessary

That it's "not super hard" to do is a real bonus.  Though I am an advocate for user flexibility, it's very reasonable to find the right balance with implementation effort.

**Comment by ericberman** [comment]
Pictures - of course.  The point is just to highlight some of the complexity, i.e., it's a bit of semantics that mean it's not a simple set of interchangeable data attributes; some are treated differently than others.
I think redundant storage would introduce more issues than it would solve.  It's not a problem if some data comes from properties vs. coming from the flight table itself.  Again, it's just an example of the nuances and complexities here.
Computed properties - given that many of the printed layouts have these, I'm certain that it would be the next request.  I've got them in the flight object precisely to support the printed layout, but again it's just one of those complexities.  

BTW, here are a few more examples of "computed" properties that I already have today:
 * Landings - I show things like "5 (2D, 1N)" which summarizes 3 data points (5 total times wheel touch down, 2 were day full-stop landings, 1 night full-stop)
 * Aircraft - I display the tail number, but you can hover for images, model, category/class, and attributes like complex or retract
 * If you have both a tach start and tach stop, I throw in total (elapsed) tach.  Ditto hobbs, engine, flight, or block.

Again, all of this is solvable, but it increases complexity, and complexity is where bugs and weird behaviors and weird limitations lurk.

**Comment by KayRJay** [comment]
> Pictures - of course. The point is just to highlight some of the complexity

Understood. 

> I think redundant storage would introduce more issues than it would solve. 

Probably true.

> Computed properties - given that many of the printed layouts have these, I'm certain that it would be the next request.

Well they are not presently in the logbook, so maybe it wouldn’t be the next request!  ;-) And maybe having them in the printed report or in flight Details is enough.  

>Again, all of this is solvable, but it increases complexity, and complexity is where bugs and weird behaviors and weird limitations lurk.

Agreed, complexity isn’t good for users or the implementation.   But often there are approaches to manage or eliminate the complexity. This is especially true if the hard things are (already) adressed or are addressable in other areas of the system (like flight Details).

I think you’ve made the argument that an 80% solution is better than no solution at all.    Don’t do the hard or complex things is the easy/easier things provide a good solution for many users.  (Or maybe it was I who made that argument!)
